src/fe-gtk/xtext.c

```diff
@@ -474,27 +474,29 @@ static void
 gtk_xtext_adjustment_set (xtext_buffer *buf, int fire_signal)
 {
 	GtkAdjustment *adj = buf->xtext->adj;
+	GtkAllocation alloc;
 
 	if (buf->xtext->buffer == buf)
 	{
-		adj->lower = 0;
-		adj->upper = buf->num_lines;
+		gtk_adjustment_set_lower (adj, 0);
+		gtk_adjustment_set_upper (adj, buf->num_lines);
 
-		if (adj->upper == 0)
-			adj->upper = 1;
+		if (gtk_adjustment_get_upper (adj) == 0)
+			gtk_adjustment_set_upper (adj, 0);
 
-		adj->page_size = GTK_WIDGET (buf->xtext)->allocation.height /
-							  buf->xtext->fontsize;
-		adj->page_increment = adj->page_size;
+		gtk_widget_get_allocation (GTK_WIDGET (buf->xtext), &alloc);
 
-		if (adj->value > adj->upper - adj->page_size)
+		gtk_adjustment_set_page_size (adj, alloc.height / buf->xtext->fontsize);
+		gtk_adjustment_set_page_increment (adj, gtk_adjustment_get_page_size (adj));
+
+		if (gtk_adjustment_get_value (adj) > gtk_adjustment_get_upper (adj) - gtk_adjustment_get_page_size (adj))
 		{
 			buf->scrollbar_down = TRUE;
-			adj->value = adj->upper - adj->page_size;
+			gtk_adjustment_set_value (adj, gtk_adjustment_get_upper (adj) - gtk_adjustment_get_page_size (adj));
```

Comment:
All of these adjustment change also trigger the change signals.

See the `fire_signal` value we track here and `gtk_adjustment_changed()` calls.